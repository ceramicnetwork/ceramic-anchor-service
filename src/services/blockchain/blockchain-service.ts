import CID from 'cids';
import Transaction from '../../models/transaction';
import Contextual from '../../contextual';

export default interface BlockchainService extends Contextual {
  /**
   * Connects to specific blockchain
   */
  connect(): Promise<void>;

  /**
   * Sends transaction with root CID as data
   */
  sendTransaction(rootCid: CID): Promise<Transaction>;

  /**
   * A string representing the CAIP-2 ID of the configured blockchain.
   */
  chainId: string;
}
