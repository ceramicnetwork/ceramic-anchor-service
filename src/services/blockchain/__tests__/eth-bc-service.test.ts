import CID from 'cids';
import Ganache from 'ganache-core'

import { config } from 'node-config-ts';

import Context from "../../../context";
import { BlockchainService } from "../blockchain-service";
import { Logger as logger } from '@overnightjs/logger/lib/Logger';

let ctx: Context = null;
let ganacheServer: any = null;
let ethBc: BlockchainService = null;

describe('ETH service',  () => {
  jest.setTimeout(25000);
  beforeAll(async () => {
    ctx = new Context();

    await ctx.build('services', 'repositories');
    ethBc = ctx.getSelectedBlockchainService();

    ganacheServer = Ganache.server({
      gasLimit: 7000000,
      time: new Date(1586784002855),
      mnemonic: 'move sense much taxi wave hurry recall stairs thank brother nut woman',
      default_balance_ether: 100,
      debug: true,
      blockTime: 2,
      network_id: 1337,
      networkId: 1337,
    });

    const localPort = config.blockchain.connectors.ethereum.rpc.port;
    const done = new Promise((resolve, reject) => {
      ganacheServer.listen(localPort, async (err: Error) => {
        if (err) {
          reject(err);
          return
        }
        resolve()
      });
    });
    await done
  });

  test('should connect to local ganache server', async () => {
    try {
      await ethBc.connect();
    } catch (err) {
      expect(false).toBe(true);
    }
  });

  test('should send CID to local ganache server', async () => {
    try {
      const cid = new CID('bafyreic5p7grucmzx363ayxgoywb6d4qf5zjxgbqjixpkokbf5jtmdj5ni');
      const tx = await ethBc.sendTransaction(cid);
      expect(tx).toBeDefined();
      expect(tx).toMatchSnapshot();
    } catch (err) {
      expect(false).toBe(true);
    }
  });

  afterAll(async (done) => {
    logger.Imp(`Closing local Ethereum blockchain instance...`);
    ganacheServer.close();
    done();
  });

});
