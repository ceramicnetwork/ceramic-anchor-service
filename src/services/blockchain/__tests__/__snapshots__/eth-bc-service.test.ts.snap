// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ETH service connected to ganache should send CID to local ganache server 1`] = `
Transaction {
  "blockNumber": 1,
  "chain": "eip155:1337",
  "txHash": "0xe72f27bb2819a17941f7f6586d118660cbabab1ebd929e810e609ded90568f55",
}
`;

exports[`ETH service with mock wallet build transaction request 1`] = `
Object {
  "data": "0x0f017112205d7fcd1a0999befdb062e6762c1f0f902f729b98304a2ef539412f53360d3d6a",
  "nonce": 5,
  "to": "abcd1234",
}
`;

exports[`ETH service with mock wallet insufficient funds error 1`] = `
Object {
  "data": "0x0f017112205d7fcd1a0999befdb062e6762c1f0f902f729b98304a2ef539412f53360d3d6a",
  "gasLimit": Object {
    "hex": "0x2710",
    "type": "BigNumber",
  },
  "gasPrice": Object {
    "hex": "0x044c",
    "type": "BigNumber",
  },
  "nonce": 5,
  "to": "abcd1234",
}
`;

exports[`ETH service with mock wallet insufficient funds error 2`] = `
Object {
  "data": "0x0f017112205d7fcd1a0999befdb062e6762c1f0f902f729b98304a2ef539412f53360d3d6a",
  "gasLimit": Object {
    "hex": "0x2710",
    "type": "BigNumber",
  },
  "gasPrice": Object {
    "hex": "0x044c",
    "type": "BigNumber",
  },
  "nonce": 5,
  "to": "abcd1234",
}
`;

exports[`ETH service with mock wallet single transaction attempt 1`] = `
Transaction {
  "blockNumber": 54321,
  "blockTimestamp": 54321000,
  "chain": "eip155:1337",
  "txHash": "0x12345abcde",
}
`;

exports[`ETH service with mock wallet single transaction attempt 2`] = `
Object {
  "data": "0x987654321",
  "gasLimit": Object {
    "hex": "0x2710",
    "type": "BigNumber",
  },
  "gasPrice": Object {
    "hex": "0x03e8",
    "type": "BigNumber",
  },
  "nonce": 5,
  "to": "abcd1234",
}
`;

exports[`ETH service with mock wallet successful mocked transaction 1`] = `
Object {
  "data": "0x0f017112205d7fcd1a0999befdb062e6762c1f0f902f729b98304a2ef539412f53360d3d6a",
  "gasLimit": Object {
    "hex": "0x2710",
    "type": "BigNumber",
  },
  "gasPrice": Object {
    "hex": "0x03e8",
    "type": "BigNumber",
  },
  "nonce": 5,
  "to": "abcd1234",
}
`;

exports[`ETH service with mock wallet timeout error 1`] = `
Object {
  "data": "0x0f017112205d7fcd1a0999befdb062e6762c1f0f902f729b98304a2ef539412f53360d3d6a",
  "gasLimit": Object {
    "hex": "0x2710",
    "type": "BigNumber",
  },
  "gasPrice": Object {
    "hex": "0x04b0",
    "type": "BigNumber",
  },
  "nonce": 5,
  "to": "abcd1234",
}
`;

exports[`ETH service with mock wallet timeout error 2`] = `
Object {
  "data": "0x0f017112205d7fcd1a0999befdb062e6762c1f0f902f729b98304a2ef539412f53360d3d6a",
  "gasLimit": Object {
    "hex": "0x2710",
    "type": "BigNumber",
  },
  "gasPrice": Object {
    "hex": "0x04b0",
    "type": "BigNumber",
  },
  "nonce": 5,
  "to": "abcd1234",
}
`;

exports[`ETH service with mock wallet timeout error 3`] = `
Object {
  "data": "0x0f017112205d7fcd1a0999befdb062e6762c1f0f902f729b98304a2ef539412f53360d3d6a",
  "gasLimit": Object {
    "hex": "0x2710",
    "type": "BigNumber",
  },
  "gasPrice": Object {
    "hex": "0x04b0",
    "type": "BigNumber",
  },
  "nonce": 5,
  "to": "abcd1234",
}
`;
